service/create/asg_stack.go

```diff
@@ -0,0 +1,222 @@
+	cloudconfig "github.com/giantswarm/k8scloudconfig"
+	"github.com/giantswarm/microerror"
+
+	awsutil "github.com/giantswarm/aws-operator/client/aws"
+	"github.com/giantswarm/aws-operator/resources"
+	awsresources "github.com/giantswarm/aws-operator/resources/aws"
+	"github.com/giantswarm/aws-operator/service/key"
+)
+
+type asgStackInput struct {
+
+	// Settings.
+	asgSize                int
+	asgType                string
+	availabilityZone       string
+	bucket                 resources.ReusableResource
+	cluster                awstpr.CustomObject
+	clusterID              string
+	iamInstanceProfileName string
+	imageID                string
+	instanceType           string
+	keyPairName            string
+	loadBalancerName       string
+	publicIP               bool
+	subnetID               string
+	tlsAssets              *certificatetpr.CompactTLSAssets
+	vpcID                  string
+	workersSecurityGroupID string
+
+	// Dependencies.
+	clients awsutil.Clients
+}
+
+type blockDeviceMapping struct {
+	DeviceName          string
+	DeleteOnTermination bool
+	VolumeSize          int64
+	VolumeType          string
+}
+
+type asgTemplateConfig struct {
+	ASGType             string
+	BlockDeviceMappings []blockDeviceMapping
+}
+
+const (
+	defaultEBSVolumeMountPoint = "/dev/xvdh"
+	// defaultEBSVolumeSize is expressed in GB.
+	defaultEBSVolumeSize = 50
+	defaultEBSVolumeType = "gp2"
+)
+
+// createASGStack creates a CloudFormation stack for an Auto Scaling Group.
+func (s *Service) createASGStack(input asgStackInput) (bool, error) {
+	var (
+		extension cloudconfig.Extension
+	)
+
+	// Generate the Cloud Formation template using a Go template.
+	cfTemplate, err := ioutil.ReadFile("resources/templates/cloudformation/auto_scaling_group.yaml")
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	goTemplate, err := template.New("asg").Parse(string(cfTemplate))
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	parsedTemplate := new(bytes.Buffer)
+	tc := asgTemplateConfig{
+		ASGType: input.asgType,
+		BlockDeviceMappings: []blockDeviceMapping{
+			{
+				DeviceName:          defaultEBSVolumeMountPoint,
+				DeleteOnTermination: true,
+				VolumeSize:          defaultEBSVolumeSize,
+				VolumeType:          defaultEBSVolumeType,
+			},
+		},
+	}
+
+	if err := goTemplate.Execute(parsedTemplate, tc); err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	// Upload the Cloud Formation template to the S3 bucket.
+	templateRelativePath := fmt.Sprintf("templates/%s.yaml", input.asgType)
+	templateURL := s.bucketObjectURL(input.cluster, templateRelativePath)
+
+	templateS3 := &awsresources.BucketObject{
+		AWSEntity: awsresources.AWSEntity{Clients: input.clients},
+		Name:      templateRelativePath,
+		Data:      parsedTemplate.String(),
+		Bucket:    input.bucket.(*awsresources.Bucket),
+	}
+	if err := templateS3.CreateOrFail(); err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	switch input.asgType {
```

Comment:
It also generates the small and large cloudconfigs. Currently this is done in the launch config which will be replaced by the ASG stack.